core/google/cloud/exceptions.py

```diff
@@ -244,8 +254,50 @@ def _walk_subclasses(klass):
             yield subsub
 
 
+@contextlib.contextmanager
+def _catch_remap_gax_error():
+    """Remap GAX exceptions that happen in context.
+
+    .. _code.proto: https://github.com/googleapis/googleapis/blob/\
+                    master/google/rpc/code.proto
+
+    Remaps gRPC exceptions to the classes defined in
+    :mod:`~google.cloud.exceptions` (according to the description
+    in `code.proto`_).
+    """
+    try:
+        yield
+    except GaxError as exc:
+        error_code = exc_to_code(exc.cause)
+        error_class = _GRPC_ERROR_MAPPING.get(error_code)
+        if error_class is None:
+            raise
+        else:
+            new_exc = error_class(exc.cause.details())
+            six.reraise(error_class, new_exc, sys.exc_info()[2])
```

Comment:
> Isn't the `exc_info` argument optional?

The [`six.reraise` docs](https://pythonhosted.org/six/#six.reraise) aren't clear what happens when one passes a new class / instance but no traceback, so I chose to be explicit (it is an error case, so performance isn't really an issue).

> Also, WDYT about six.raise_from here instead?

AFAICT, [`six.raise_from`](https://pythonhosted.org/six/#six.raise_from) *would* discard the existing traceback on Python2.